File: src/lib/sorteo.ts
/**
 * Algoritmo de sorteo para asignación de amigos secretos
 * Garantiza que nadie se toque a sí mismo y no hay intercambios equivalentes
 */

interface Participant {
  id: string;
  category: string;
}

/**
 * Algoritmo de Shuffle (Fisher-Yates)
 */
function shuffle<T>(array: T[]): T[] {
  const result = [...array];
  for (let i = result.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [result[i], result[j]] = [result[j], result[i]];
  }
  return result;
}

/**
 * Valida que la asignación cumpla las restricciones
 */
function isValidAssignment(
  participants: Participant[],
  assignments: Map<string, string>
): boolean {
  for (const participant of participants) {
    const assignedTo = assignments.get(participant.id);
    
    // No puede tocarse a sí mismo
    if (assignedTo === participant.id) {
      return false;
    }
    
    // No puede haber intercambios equivalentes (si A->B, entonces B no puede ser ->A)
    const assignedToTarget = assignments.get(assignedTo!);
    if (assignedToTarget === participant.id) {
      return false;
    }
  }
  
  return true;
}

/**
 * Crea una asignación circular válida
 * Evita ciclos de 2 (A->B, B->A) garantizando un ciclo más largo
 */
function createValidAssignment(participants: Participant[]): Map<string, string> {
  const n = participants.length;
  
  if (n < 2) {
    throw new Error('Se necesitan al menos 2 participantes para el sorteo');
  }
  
  // Para evitar ciclos de 2, creamos un ciclo único que incluye a todos
  // Esto garantiza que si A->B, B nunca apuntará a A directamente
  let attempts = 0;
  const maxAttempts = 1000;
  
  while (attempts < maxAttempts) {
    const shuffled = shuffle(participants);
    const assignments = new Map<string, string>();
    
    // Crear un ciclo: cada persona apunta a la siguiente, la última apunta a la primera
    for (let i = 0; i < n; i++) {
      const nextIndex = (i + 1) % n;
      assignments.set(shuffled[i].id, shuffled[nextIndex].id);
    }
    
    // Validar que no hay ciclos de 2 (esto ya está garantizado por la construcción del ciclo)
    // Pero verificamos por si acaso
    if (isValidAssignment(participants, assignments)) {
      return assignments;
    }
    
    attempts++;
  }
  
  throw new Error('No se pudo generar una asignación válida después de múltiples intentos');
}

/**
 * Realiza el sorteo por categoría
 */
export function performSorteo(
  eliteParticipants: Participant[],
  diversionParticipants: Participant[]
): Map<string, string> {
  const assignments = new Map<string, string>();
  
  // Sorteo para categoría Élite
  if (eliteParticipants.length >= 2) {
    const eliteAssignments = createValidAssignment(eliteParticipants);
    eliteAssignments.forEach((value, key) => assignments.set(key, value));
  } else if (eliteParticipants.length === 1) {
    console.warn('Solo hay 1 participante en categoría Élite, no se puede asignar');
  }
  
  // Sorteo para categoría Diversión
  if (diversionParticipants.length >= 2) {
    const diversionAssignments = createValidAssignment(diversionParticipants);
    diversionAssignments.forEach((value, key) => assignments.set(key, value));
  } else if (diversionParticipants.length === 1) {
    console.warn('Solo hay 1 participante en categoría Diversión, no se puede asignar');
  }
  
  return assignments;
}

/**
 * Verifica que todas las asignaciones sean válidas
 */
export function validateAssignments(
  participants: Participant[],
  assignments: Map<string, string>
): { valid: boolean; errors: string[] } {
  const errors: string[] = [];
  const assignedTo = new Set<string>();
  
  for (const participant of participants) {
    const target = assignments.get(participant.id);
    
    if (!target) {
      errors.push(`Participante ${participant.id} no tiene asignación`);
      continue;
    }
    
    // Verificar que no se asigne a sí mismo
    if (target === participant.id) {
      errors.push(`Participante ${participant.id} está asignado a sí mismo`);
    }
    
    // Verificar que el objetivo existe
    const targetExists = participants.find(p => p.id === target);
    if (!targetExists) {
      errors.push(`El objetivo ${target} no existe`);
    }
    
    // Verificar categorías coinciden
    if (targetExists && targetExists.category !== participant.category) {
      errors.push(`Participante ${participant.id} y su objetivo ${target} tienen diferentes categorías`);
    }
    
    assignedTo.add(target);
  }
  
  // Verificar que todos reciban regalo (en su categoría)
  const categoryCounts = new Map<string, { total: number; receiving: number }>();
  
  for (const participant of participants) {
    const category = participant.category;
    if (!categoryCounts.has(category)) {
      categoryCounts.set(category, { total: 0, receiving: 0 });
    }
    const counts = categoryCounts.get(category)!;
    counts.total++;
    
    if (assignedTo.has(participant.id)) {
      counts.receiving++;
    }
  }
  
  categoryCounts.forEach((counts, category) => {
    if (counts.total !== counts.receiving && counts.total > 1) {
      errors.push(`En categoría ${category}: ${counts.total} participantes pero solo ${counts.receiving} recibirán regalo`);
    }
  });
  
  return {
    valid: errors.length === 0,
    errors
  };
}

File: lib/sorteo.py
"""
Algoritmo de sorteo para asignación de amigos secretos
Garantiza que nadie se toque a sí mismo y no hay intercambios equivalentes
Portado desde TypeScript (sorteo.ts)
"""

import random
from typing import List, Dict, Tuple


def shuffle(array: list) -> list:
    """
    Algoritmo de Fisher-Yates para mezclar aleatoriamente un array
    
    Args:
        array: Lista a mezclar
        
    Returns:
        Lista mezclada
    """
    result = array.copy()
    for i in range(len(result) - 1, 0, -1):
        j = random.randint(0, i)
        result[i], result[j] = result[j], result[i]
    return result


def is_valid_assignment(participants: List[dict], assignments: Dict[str, str]) -> bool:
    """
    Valida que la asignación cumpla las restricciones:
    - Nadie se toca a sí mismo
    - No hay intercambios equivalentes (si A->B, entonces B no puede ser ->A)
    
    Args:
        participants: Lista de participantes con id y category
        assignments: Diccionario {participant_id: assigned_to_id}
        
    Returns:
        True si la asignación es válida
    """
    for participant in participants:
        participant_id = participant['id']
        assigned_to = assignments.get(participant_id)
        
        # No puede tocarse a sí mismo
        if assigned_to == participant_id:
            return False
        
        # No puede haber intercambios equivalentes
        assigned_to_target = assignments.get(assigned_to)
        if assigned_to_target == participant_id:
            return False
    
    return True


def create_valid_assignment(participants: List[dict]) -> Dict[str, str]:
    """
    Crea una asignación circular válida para un grupo de participantes
    Evita ciclos de 2 (A->B, B->A) garantizando un ciclo más largo
    
    Args:
        participants: Lista de participantes con id y category
        
    Returns:
        Diccionario con las asignaciones {participant_id: assigned_to_id}
        
    Raises:
        Exception: Si no se puede generar una asignación válida
    """
    n = len(participants)
    
    if n < 2:
        raise Exception('Se necesitan al menos 2 participantes para el sorteo')
    
    # Para evitar ciclos de 2, creamos un ciclo único que incluye a todos
    # Esto garantiza que si A->B, B nunca apuntará a A directamente
    max_attempts = 1000
    
    for attempt in range(max_attempts):
        shuffled = shuffle(participants)
        assignments = {}
        
        # Crear un ciclo: cada persona apunta a la siguiente, la última apunta a la primera
        for i in range(n):
            next_index = (i + 1) % n
            assignments[shuffled[i]['id']] = shuffled[next_index]['id']
        
        # Validar que no hay ciclos de 2
        if is_valid_assignment(participants, assignments):
            return assignments
    
    raise Exception('No se pudo generar una asignación válida después de múltiples intentos')


def perform_sorteo(elite_participants: List[dict], diversion_participants: List[dict]) -> Dict[str, str]:
    """
    Realiza el sorteo por categoría
    
    Args:
        elite_participants: Lista de participantes de categoría élite
        diversion_participants: Lista de participantes de categoría diversión
        
    Returns:
        Diccionario con todas las asignaciones {participant_id: assigned_to_id}
        
    Raises:
        Exception: Si hay solo 1 participante en alguna categoría
    """
    assignments = {}
    
    # Sorteo para categoría Élite
    if len(elite_participants) >= 2:
        elite_assignments = create_valid_assignment(elite_participants)
        assignments.update(elite_assignments)
    elif len(elite_participants) == 1:
        raise Exception('Solo hay 1 participante en categoría Élite, no se puede asignar')
    
    # Sorteo para categoría Diversión
    if len(diversion_participants) >= 2:
        diversion_assignments = create_valid_assignment(diversion_participants)
        assignments.update(diversion_assignments)
    elif len(diversion_participants) == 1:
        raise Exception('Solo hay 1 participante en categoría Diversión, no se puede asignar')
    
    return assignments


def validate_assignments(participants: List[dict], assignments: Dict[str, str]) -> dict:
    """
    Valida las asignaciones del sorteo y retorna un reporte
    
    Args:
        participants: Lista de todos los participantes
        assignments: Diccionario con las asignaciones
        
    Returns:
        Dict con 'valid' (bool), 'errors' (List[str]) y 'summary' (str)
    """
    errors = []
    
    # Verificar que todos tengan asignación
    for participant in participants:
        if participant['id'] not in assignments:
            errors.append(f"Participante {participant['id']} no tiene asignación")
    
    # Verificar que nadie se toque a sí mismo
    for participant_id, assigned_to_id in assignments.items():
        if participant_id == assigned_to_id:
            errors.append(f"Participante {participant_id} se asignó a sí mismo")
    
    # Verificar que no haya intercambios equivalentes
    for participant_id, assigned_to_id in assignments.items():
        if assignments.get(assigned_to_id) == participant_id:
            errors.append(f"Intercambio equivalente detectado: {participant_id} <-> {assigned_to_id}")
    
    # Verificar que todos los assigned_to_id sean participantes válidos
    participant_ids = {p['id'] for p in participants}
    for participant_id, assigned_to_id in assignments.items():
        if assigned_to_id not in participant_ids:
            errors.append(f"Asignación inválida: {participant_id} -> {assigned_to_id} (no existe)")
    
    # Verificar asignaciones por categoría
    participants_by_id = {p['id']: p for p in participants}
    for participant_id, assigned_to_id in assignments.items():
        if participant_id in participants_by_id and assigned_to_id in participants_by_id:
            if participants_by_id[participant_id]['category'] != participants_by_id[assigned_to_id]['category']:
                errors.append(
                    f"Asignación entre categorías diferentes: {participant_id} ({participants_by_id[participant_id]['category']}) -> "
                    f"{assigned_to_id} ({participants_by_id[assigned_to_id]['category']})"
                )
    
    valid = len(errors) == 0
    summary = f"Validación {'exitosa' if valid else 'fallida'}: {len(assignments)} asignaciones"
    
    return {
        'valid': valid,
        'errors': errors,
        'summary': summary
    }
